### ISSUE TITLE
API requests to edge router failing

### ISSUE THREAD

**Issue by sndrsnk** [localstack/localstack#2703]
# Bug Report

# Detailed description
I'm using serverless to create some lambdas that can be triggered via API calls, and I use serverless-localstack (with LocalStack set up as a separate container via (docker-compose.yml) to which I'm deploying my stack. 

However, when I make a simple cURL request to `http://localhost:4566/hello` from the host I receive the following error: 

```nohighlight
2020-07-09T10:42:13:INFO:localstack.services.edge: Unable to find forwarding rule for host "localhost:4566", path "/hello", target header "", auth header "", data "b''"
```

Response: `{"status": "running"}`. 

It works as intended when I make the same request directly to `http://localhost:4567/restapis/g2grzvpfhe/local/_user_request_/hello`. 

## Expected behavior
I expect to see API response data from calling `http://localhost:4566/hello` directly. 

## Actual behavior
I receive a `{"status": "running"}` response with the above error reported in the Docker logs. 

# Steps to reproduce

## Command used to start LocalStack
`./node_modules/serverless/bin/serverless.js invoke -s local -f LambdaTest -p ./tests/resources/base_event.yml`

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
N/A

Note: I'm using docker image `localstack/localstack:latest`.

**Comment by whummer** [comment]
Thanks for reporting @sndrsnk . We need to understand the desired behavior here a bit better, given that the edge service at `localhost:4566` is used to serve all sorts of different API requests. We'd have to agree on a reasonable API lookup approach that does not rely on domain names and is able to expose uniquely identifiable API endpoints on the central edge port, without overwriting existing routes that are already defined for other LocalStack services or APIs.

Looking at the API Gateway docs [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-deploy-api.html) there are two ways of invoking an API in real AWS.

(1) Standard format is:
```
https://{restapi-id}.execute-api.{region}.amazonaws.com/{stageName}/...
```
(2) Alternatively, if a custom domain name is associated with the API:
```
https://{api.example.com}/{basePath}
```

The example you've posted is the standard way of invoking APIs (it maps the details from (1) from domain based to path based addressing):
```
http://localhost:4567/restapis/g2grzvpfhe/local/_user_request_/hello
```

In the example you proposed (posting to `http://localhost:4566/hello`), would it be feasible for you to specify a `Host: api.example.com` HTTP header with the custom domain name of the API? This would help us more uniquely identify the target API, and avoid any collisions with other existing edge route mappings. Thanks

**Comment by phillip-haydon** [comment]
Isn't the issue that the documentation says to use `localhost:4566` but it doesn't work at all?

I just tried out localstack today for the first time, went to dockerhub `https://hub.docker.com/r/localstack/localstack` and created a docker-compose file, copied the config in, ran `docker-compose up` and when I navigate to `http://localstack:4566/` i get nothing. 

So I figured ok I guess there's no dashboard or anything like that... So I tried creating a bucket.

> phillip  pop-os  ~  $  aws --endpoint-url=http://localhost:4566/ s3 mb s3://test
> make_bucket failed: s3://test Could not connect to the endpoint URL: "http://localhost:4566/test"
> phillip  pop-os  ~  $  aws --endpoint-url=https://localhost:4566/ s3 mb s3://test
> make_bucket failed: s3://test Could not connect to the endpoint URL: "https://localhost:4566/test"
> phillip  pop-os  ~  ERROR  $  aws --endpoint-url=http://localhost:4572/ s3 mb s3://test
> make_bucket: test

So I'm a bit confused with how localstack is meant to work at the moment. :(

**Comment by whummer** [comment]
Thanks for reporting @phillip-haydon . Can you please share your docker-compose configuration - is it the one from the repository? Please ensure that port 4566 is exposed for the container.

I just did a quick test with the `latest` image and the docker-compose from the repository - seems to be working:
```
$ curl localhost:4566
{"status": "running"}
```
Please also share the log output from the Docker container. Thanks for your help.

**Comment by phillip-haydon** [comment]
@whummer Thank you!!!!

So I was using the one defined on dockerhub. 

https://hub.docker.com/r/localstack/localstack

This didn't work. I swapped to the one in the repo, changing only the image to `localstack/localstack:latest` cos I had 0.11.3 running previously.

It spins up fine and works without issue now. 

![image](https://user-images.githubusercontent.com/895629/87510099-84509b00-c6a5-11ea-80ca-5ebcad60279e.png)

**Comment by sndrsnk** [comment]
@whummer Thanks for getting back to me. You mentioned that it's possible to have a hostname in place of `http://localhost:4567/restapis/g2grzvpfhe/local/_user_request_/` for instance -- it may not be the place to ask but is this something that is achievable with the LocalStack/Serverless plugin whereby a hostname can be created locally after the deployment of a stack to LocalStack? 

One of our issues is that we're having difficulty with the unique identifier of the API endpoint, i.e. `g2grzvpfhe`, which makes it difficult to perform API endpoint tests locally.

**Comment by whummer** [comment]
Thanks, @sndrsnk .

> it may not be the place to ask but is this something that is achievable with the LocalStack/Serverless plugin whereby a hostname can be created locally after the deployment of a stack to LocalStack?

It is possible, yes - if you're on Mac/Linux, you could adding an entry to `/etc/hosts` to make your system resolve the host name.

The original issue you reported should now also be fixed via #2748 - the `.../_user_request_/...` URLs are now also accessible through the edge service on port `4566`. Closing this issue - but please report here if the problem persists or if there are any other issues related to this feature. Thanks
